### PR TITLE
Add memory sanitizer (detects e.g. uninitialized use)

### PR DESCRIPTION
**Pull request by bjodah**

### CODE REVIEW

**Comment by bjodah** [comment]
Turns out one needs an instrumented `libc++` for this to work:
https://github.com/google/sanitizers/wiki/MemorySanitizerLibcxxHowTo#instrumented-libc

Compiling that during CI runs isn't really practical. Perhaps having an `msan` enabled `libc++` on `conda-forge` would make it feasible...

**Comment by isuruf** [comment]
Can we make use of libc++ from ubuntu repo? Currently we are using libstdc++. (We'll have to test only with C dependencies as C++ dependencies like LLVM libraries are compiled with libstdc++)
Btw, libc++ is really small and takes around a minute to compile.

File: bin/test_travis.sh [hunk 1]
```diff
@@ -104,6 +104,15 @@ if [[ "${WITH_SANITIZE}" != "" ]]; then
...
 	    export UBSAN_OPTIONS=print_stacktrace=1,halt_on_error=1,external_symbolizer_path=/usr/lib/llvm-7/bin/llvm-symbolizer
+	elif [[ "${WITH_SANITIZE}" == "memory" ]]; then
+	    export MSAN_OPTIONS=halt_on_error=1,external_symbolizer_path=/usr/lib/llvm-7/bin/llvm-symbolizer
+            LLVM_ORG_VER=${WITH_LLVM}.1  # 7.0.1, using ${WITH_LLVM} so we rembmer to update this
+            curl -Ls https://github.com/llvm/llvm-project/archive/llvmorg-${LLVM_ORG_VER}.tar.gz | tar xz -C /tmp
+            mkdir /opt/libcxx_msan && cd /opt/libcxx_msan
+            cmake -DCMAKE_BUILD_TYPE=Release -DLLVM_USE_SANITIZER=Memory -DCMAKE_C_COMPILER=clang -DCMAKE_CXX_COMPILER=clang++ /tmp/llvmorg-${LLVM_ORG_VER}/libcxx
+            make cxx
+            export CXXFLAGS="$CXXFLAGS -stdlib=libc++ -L/opt/libcxx_msan/lib -lc++abi -I/opt/libcxx_msan/include -I/opt/libcxx_msan/include/c++/v1"
+            export LDFLAGS="$LDFLAGS -Wl,-rpath,/opt/libcxx_msan/lib"
```
Comment (bjodah) [inline bin/test_travis.sh]: me setting LDFLAGS here won't help right? Do I need to set `CMAKE_SHARED_LINKER_FLAGS`?
perhaps this:
`cmake_line=$cmake_line -DCMAKE_SHARED_LINKER_FLAGS="-Wl,-rpath,/opt/libcxx_msan/lib"`

Comment (isuruf) [reply bin/test_travis.sh]: Setting LDFLAGS is fine. They'll be picked up by cmake.

**Comment by bjodah** [comment]
Reading up more on memory sanitizer I see this [over at Mozilla](https://developer.mozilla.org/en-US/docs/Mozilla/Memory_Sanitizer#Building_Firefox):

> Warning: Running Firefox with MemorySanitizer would require all external dependencies to be built with MemorySanitizer as well. To our knowledge, this has never been attempted yet, so the build configuration provided here is untested and without an appropriately instrumented userland, it will cause false positives.

and then [further down on the same page](https://developer.mozilla.org/en-US/docs/Mozilla/Memory_Sanitizer#Building_the_JavaScript_shell):

> Note: Unlike Firefox itself, the JavaScript shell does not require an instrumented userland. Calls to external libraries like zlib are handled with special annotations inside the engine.

So maybe we'll need to build using `-DWITH_LLVM=OFF` and `-DINTEGER_CLASS=boostmp` or so for the msan build...

**Comment by isuruf** [comment]
```
So maybe we'll need to build using -DWITH_LLVM=OFF and -DINTEGER_CLASS=boostmp or so for the msan build...
```

Good idea. We can first try `-DINTEGER_CLASS=gmp` and if that doesn't work, move to `boostmp`

**Comment by isuruf** [comment]
From the introduction in the wiki page you mentioned

> If you want MemorySanitizer to work properly and not produce any false positives, you must ensure that all the code in your program and in libraries it uses is instrumented

Let's use `INTEGER_CLASS=boostmp` for msan build

**Comment by bjodah** [comment]
I think we need to call cmak & make for the libcxxabi folder as well (it was missing in the instructions on their wiki).

**Comment by bjodah** [comment]
I re-added the `-lc++abi` flag due to errors like:
```
 /opt/libcxx_msan/lib/libc++.so: undefined reference to `vtable for
  __cxxabiv1::__si_class_type_info'
```

**Comment by isuruf** [comment]
That error might be because the build is picking up a `clang-5` from somewhere else

**Comment by bjodah** [comment]
I noticed it's doing that (which is weird since it's not for the other sanitizers), should we just set `CXX=clang++-7 CC=clang-7`?

File: .travis.yml [hunk 2]
```diff
@@ -170,7 +170,7 @@ matrix:
           packages:
           - clang-7
           - llvm-7-dev
-    - env: BUILD_TYPE="Debug" WITH_SANITIZE="memory" INTEGER_CLASS="boostmp"
+    - env: BUILD_TYPE="Debug" WITH_SANITIZE="memory" INTEGER_CLASS="boostmp" CXX=clang++-7 CC=clang-7
```
Comment (isuruf) [inline .travis.yml]: `CXX` here gets overriden by the `compiler: clang` below.

Comment (bjodah) [reply .travis.yml]: ok, but how come e.g. the run with `WITH_SANITIZE="address"` [does this](https://travis-ci.org/symengine/symengine/jobs/512764111#L2285)?

Comment (isuruf) [reply .travis.yml]: It's set at https://github.com/symengine/symengine/blob/master/bin/install_travis.sh#L106

Comment (bjodah) [reply .travis.yml]: thanks!

**Comment by bjodah** [comment]
Looks like it compiles and most tests pass, [2 exceptions](https://travis-ci.org/symengine/symengine/jobs/516944747#L2287):
```
The following tests FAILED:
	 42 - test_bipartite (OTHER_FAULT)
	 43 - test_hopcroft_karp (OTHER_FAULT)
```

**Comment by certik** [comment]
@bjodah what is the status of this? What needs to be done to finish it?

**Comment by bjodah** [comment]
My guess is that there is at least one bug in the pattern matching code (since msan fails those tests). I.e. uninitialized reads. However, I could never figure out how to get msan to print a backtrace indicating the offending line so I'm left in the dark here. The documentation for the sanitizers is quite sparse...

**Comment by bjodah** [comment]
Thanks. I squashed a bunch of commits leaving a handful (feel free to "squash-and-merge" as a single commit), it would have been nice to fix the matchcpp test failures, but if tests pass I'm good to merge.

**Comment by certik** [comment]
Do you have a link to the failures? I don't know how difficult it would be.

**Comment by bjodah** [comment]
The link in my previous comment is still valid:
https://github.com/symengine/symengine/pull/1543#issuecomment-480622094

(even if the particular commit has been rebased away)

**Comment by certik** [comment]
I have a few more questions:

* Is there a way to debug this locally? Say in Ubuntu 18.04?
* Are you installing llvm from source on Travis?

**Comment by bjodah** [comment]
I guess one can use lldb or gdb to debug it, but ideally it would print a backtrace. I'm just not familiar with the code in question.

No, we're only recompiling libcxx with `-fsanitize=memory` (since the instrumentation requires that all libraries linked against are also instrumented).

**Comment by certik** [comment]
Here is what I did on Ubuntu 18.04:
```
CXX=clang CXXFLAGS="-fsanitize=memory" cmake -DCMAKE_BUILD_TYPE=Debug .
make
```
and I got 6762 lines of error messages.....
They start with:
```
[  4%] Built target teuchos
[ 32%] Built target symengine
[ 33%] Built target catch
[ 34%] Linking CXX executable test_rcp
CMakeFiles/test_rcp.dir/test_rcp.cpp.o: In function `____C_A_T_C_H____T_E_S_T____0()':
/home/certik/repos/symengine/symengine/symengine/tests/rcp/test_rcp.cpp:23: undefined reference to `__cxa_begin_catch'
/home/certik/repos/symengine/symengine/symengine/tests/rcp/test_rcp.cpp:23: undefined reference to `__cxa_end_catch'
```

If you send me the exact commands that recompile symengine with memory sanitizer, I am happy to debug and fix the C++ code.

**Comment by bjodah** [comment]
Hmm... So locally, I seem to trigger msan on all tests (in `__dynamic_cast` so I've probably messed up somehow).

Here's a bash script which creates a Dockerimage and runs the compilation locally in Ubuntu 18.04:
https://gist.github.com/bjodah/82e1013e2025003e35f365b89921ddc2

here's [ctest's output](https://pastebin.com/3q8Dpizc)

I'm out of time for now though...

**Comment by certik** [comment]
Let's not merge until we figure this out.

**Comment by bjodah** [comment]
Ok, so I get 5 errors when building with memory sanitizer on current master:
<details><summary>ctest --output-on-failure</summary>

```
+ ctest --output-on-failure
Test project /mount/symengine-build
      Start  1: test_rcp
 1/50 Test  #1: test_rcp .........................   Passed    0.02 sec
      Start  2: test_basic
 2/50 Test  #2: test_basic .......................***Failed    0.03 sec
2*x + 3*y
2*x + 3*y
2*x + 2*y
2*x
2*x + y
----------------------
5 + 2*x*y + x**2
5
6
11
-1
30
5/6
15625
1/15625
-5
5/6

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
test_basic is a Catch v1.12.1 host application.
Run with -? for options

-------------------------------------------------------------------------------
Rational: Basic
-------------------------------------------------------------------------------
/mount/symengine-d0b1932b4ffaf4b364d4b436a796b456917e079b/symengine/tests/basic/test_basic.cpp:337
...............................................................................

/mount/symengine-d0b1932b4ffaf4b364d4b436a796b456917e079b/symengine/tests/basic/test_basic.cpp:337: FAILED:
  {Unknown expression after the reported line}
due to unexpected exception with message:
  bad rational: non-zero singular denominator

x**2*y**3
x**2*y**3
x**2*y**2
===============================================================================
test cases:  15 |  14 passed | 1 failed
assertions: 250 | 249 passed | 1 failed


      Start  3: test_arit
 3/50 Test  #3: test_arit ........................   Passed    0.09 sec
      Start  4: test_poly
 4/50 Test  #4: test_poly ........................   Passed    0.01 sec
      Start  5: test_series
 5/50 Test  #5: test_series ......................   Passed    0.01 sec
      Start  6: test_series_generic
 6/50 Test  #6: test_series_generic ..............   Passed    0.50 sec
      Start  7: test_functions
 7/50 Test  #7: test_functions ...................   Passed    0.09 sec
      Start  8: test_subs
 8/50 Test  #8: test_subs ........................   Passed    0.01 sec
      Start  9: test_integer
 9/50 Test  #9: test_integer .....................   Passed    0.01 sec
      Start 10: test_rational
10/50 Test #10: test_rational ....................   Passed    0.01 sec
      Start 11: test_relationals
11/50 Test #11: test_relationals .................   Passed    0.01 sec
      Start 12: test_number
12/50 Test #12: test_number ......................   Passed    0.01 sec
      Start 13: test_as_numer_denom
13/50 Test #13: test_as_numer_denom ..............***Failed    0.01 sec

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
test_as_numer_denom is a Catch v1.12.1 host application.
Run with -? for options

-------------------------------------------------------------------------------
NumerDenom: Rational
-------------------------------------------------------------------------------
/mount/symengine-d0b1932b4ffaf4b364d4b436a796b456917e079b/symengine/tests/basic/test_as_numer_denom.cpp:46
...............................................................................

/mount/symengine-d0b1932b4ffaf4b364d4b436a796b456917e079b/symengine/tests/basic/test_as_numer_denom.cpp:46: FAILED:
due to unexpected exception with message:
  bad rational: non-zero singular denominator

===============================================================================
test cases:  6 |  5 passed | 1 failed
assertions: 35 | 34 passed | 1 failed


      Start 14: test_parser
14/50 Test #14: test_parser ......................   Passed    0.03 sec
      Start 15: test_sets
15/50 Test #15: test_sets ........................   Passed    0.02 sec
      Start 16: test_fields
16/50 Test #16: test_fields ......................   Passed    0.49 sec
      Start 17: test_infinity
17/50 Test #17: test_infinity ....................   Passed    0.01 sec
      Start 18: test_nan
18/50 Test #18: test_nan .........................   Passed    0.01 sec
      Start 19: test_solve
19/50 Test #19: test_solve .......................   Passed    0.10 sec
      Start 20: test_as_real_imag
20/50 Test #20: test_as_real_imag ................   Passed    0.02 sec
      Start 21: test_integer_class
21/50 Test #21: test_integer_class ...............   Passed    0.01 sec
      Start 22: test_cse
22/50 Test #22: test_cse .........................   Passed    0.02 sec
      Start 23: test_count_ops
23/50 Test #23: test_count_ops ...................   Passed    0.01 sec
      Start 24: test_ntheory
24/50 Test #24: test_ntheory .....................   Passed    0.10 sec
      Start 25: test_diophantine
25/50 Test #25: test_diophantine .................   Passed    0.01 sec
      Start 26: test_matrix
26/50 Test #26: test_matrix ......................   Passed    0.05 sec
      Start 27: test_printing
27/50 Test #27: test_printing ....................   Passed    0.02 sec
      Start 28: test_ccode
28/50 Test #28: test_ccode .......................***Failed    0.01 sec

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
test_ccode is a Catch v1.12.1 host application.
Run with -? for options

-------------------------------------------------------------------------------
Rational
-------------------------------------------------------------------------------
/mount/symengine-d0b1932b4ffaf4b364d4b436a796b456917e079b/symengine/tests/printing/test_ccode.cpp:80
...............................................................................

/mount/symengine-d0b1932b4ffaf4b364d4b436a796b456917e079b/symengine/tests/printing/test_ccode.cpp:80: FAILED:
  {Unknown expression after the reported line}
due to unexpected exception with message:
  bad rational: non-zero singular denominator

===============================================================================
test cases:  7 |  6 passed | 1 failed
assertions: 49 | 48 passed | 1 failed


      Start 29: test_uintpoly
29/50 Test #29: test_uintpoly ....................   Passed    0.01 sec
      Start 30: test_uratpoly
30/50 Test #30: test_uratpoly ....................   Passed    0.02 sec
      Start 31: test_mintpoly
31/50 Test #31: test_mintpoly ....................   Passed    0.02 sec
      Start 32: test_uexprpoly
32/50 Test #32: test_uexprpoly ...................   Passed    0.02 sec
      Start 33: test_mexprpoly
33/50 Test #33: test_mexprpoly ...................   Passed    0.03 sec
      Start 34: test_basic_conversions
34/50 Test #34: test_basic_conversions ...........   Passed    0.02 sec
      Start 35: test_evalf
35/50 Test #35: test_evalf .......................   Passed    0.01 sec
      Start 36: test_eval_double
36/50 Test #36: test_eval_double .................   Passed    0.02 sec
      Start 37: test_lambda_double
37/50 Test #37: test_lambda_double ...............   Passed    0.01 sec
      Start 38: test_cwrapper
38/50 Test #38: test_cwrapper ....................   Passed    0.02 sec
      Start 39: test_expression
39/50 Test #39: test_expression ..................   Passed    0.02 sec
      Start 40: test_logic
40/50 Test #40: test_logic .......................   Passed    0.01 sec
      Start 41: test_finitediff
41/50 Test #41: test_finitediff ..................   Passed    0.01 sec
      Start 42: test_bipartite
42/50 Test #42: test_bipartite ...................***Exception: Child aborted  0.15 sec
uncaught_exceptions not yet implemented

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
test_bipartite is a Catch v1.12.1 host application.
Run with -? for options

-------------------------------------------------------------------------------
BipartiteGraph
  Test 1
-------------------------------------------------------------------------------
/mount/symengine-d0b1932b4ffaf4b364d4b436a796b456917e079b/symengine/utilities/matchpycpp/tests/bipartite/test_bipartite.cpp:9
...............................................................................

/mount/symengine-d0b1932b4ffaf4b364d4b436a796b456917e079b/symengine/utilities/matchpycpp/tests/bipartite/test_bipartite.cpp:9: FAILED:
  {Unknown expression after the reported line}
with expansion:

due to a fatal error condition:
  SIGABRT - Abort (abnormal termination) signal

===============================================================================
test cases: 1 | 1 failed
assertions: 2 | 1 passed | 1 failed


      Start 43: test_hopcroft_karp
43/50 Test #43: test_hopcroft_karp ...............***Exception: Child aborted  0.12 sec
uncaught_exceptions not yet implemented

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
test_hopcroft_karp is a Catch v1.12.1 host application.
Run with -? for options

-------------------------------------------------------------------------------
Hopcroft Karp algorithm
  Test 1
-------------------------------------------------------------------------------
/mount/symengine-d0b1932b4ffaf4b364d4b436a796b456917e079b/symengine/utilities/matchpycpp/tests/hopcroft_karp/test_hopcroft_karp.cpp:11
...............................................................................

/mount/symengine-d0b1932b4ffaf4b364d4b436a796b456917e079b/symengine/utilities/matchpycpp/tests/hopcroft_karp/test_hopcroft_karp.cpp:11: FAILED:
  {Unknown expression after the reported line}
with expansion:

due to a fatal error condition:
  SIGABRT - Abort (abnormal termination) signal

===============================================================================
test cases: 1 | 1 failed
assertions: 3 | 2 passed | 1 failed


      Start 44: test_case001
44/50 Test #44: test_case001 .....................   Passed    0.01 sec
      Start 45: test_case002
45/50 Test #45: test_case002 .....................   Passed    0.01 sec
      Start 46: test_case003
46/50 Test #46: test_case003 .....................   Passed    0.01 sec
      Start 47: test_case004
47/50 Test #47: test_case004 .....................   Passed    0.01 sec
      Start 48: test_case005
48/50 Test #48: test_case005 .....................   Passed    0.01 sec
      Start 49: test_case006
49/50 Test #49: test_case006 .....................   Passed    0.01 sec
      Start 50: test_case007
50/50 Test #50: test_case007 .....................   Passed    0.01 sec

90% tests passed, 5 tests failed out of 50

Total Test time (real) =   2.31 sec

The following tests FAILED:
	  2 - test_basic (Failed)
	 13 - test_as_numer_denom (Failed)
	 28 - test_ccode (Failed)
	 42 - test_bipartite (Child aborted)
	 43 - test_hopcroft_karp (Child aborted)
Errors while running CTest
```

</details>

bash script to reproduce (you need to have docker installed):
<details><summary>$ ./docker_symengine_msan.sh d0b1932b4ffaf4b364d4b436a796b456917e079b</summary>

```
#!/bin/bash -xeu
# This script uses docker to build a specific symengine commit with memory sanitizer enabled.
#
# Example usage:
#
#  $ ./docker_symengine_msan.sh e7d7ae55f4fd6740a48acaea0e7e24542ceda6b7
#
LLVM_ORG_VER=8.0.1

if ! which docker; then
    2>&1 echo "You need to have docker installed. See e.g. https://www.docker.com/community-edition#/download"
    exit 1
fi
if [ ! -d environment_msan ]; then
   mkdir environment_msan
   cat <<EOF >environment_msan/Dockerfile
FROM ubuntu:18.04
RUN apt-get update && \\
    apt-get --quiet --assume-yes --no-install-recommends install libgmp-dev binutils-dev cmake curl zlib1g-dev sudo wget gnupg2 make && \
    wget --no-check-certificate -O - https://apt.llvm.org/llvm-snapshot.gpg.key|sudo apt-key add - && \
    echo "deb http://apt.llvm.org/bionic/ llvm-toolchain-bionic-8 main" | tee -a /etc/apt/sources.list && \
    apt-get update && apt-get --quiet --assume-yes --no-install-recommends install clang-8 llvm-8-dev
ENV CC=clang-8 CXX=clang++-8
RUN cd /tmp
RUN curl -Lsk https://github.com/llvm/llvm-project/archive/llvmorg-${LLVM_ORG_VER}.tar.gz | tar xz -C /tmp && \
    mkdir /tmp/build_libcxx8 &&  \
    cd /tmp/build_libcxx8 &&  \
    cmake -DCMAKE_BUILD_TYPE=Release -DLLVM_USE_SANITIZER=Memory -DLLVM_CONFIG_PATH=/usr/bin/llvm-config-8 -DCMAKE_INSTALL_PREFIX=/opt/libcxx8_msan /tmp/llvm-project-llvmorg-${LLVM_ORG_VER}/libcxx &&  \
    cmake --build .  && \
    cmake --build . --target install && \
    mkdir /tmp/build_libcxxabi8 &&  \
    cd /tmp/build_libcxxabi8 &&  \
    cmake -DCMAKE_BUILD_TYPE=Release -DLLVM_USE_SANITIZER=Memory -DLLVM_CONFIG_PATH=/usr/bin/llvm-config-8 -DCMAKE_INSTALL_PREFIX=/opt/libcxx8_msan -DLIBCXXABI_LIBCXX_INCLUDES=/opt/libcxx8_msan/include/c++/v1 -DLIBCXXABI_LIBCXX_PATH=/tmp/llvm-project-llvmorg-${LLVM_ORG_VER}/libcxx  /tmp/llvm-project-llvmorg-${LLVM_ORG_VER}/libcxxabi &&  \
    cmake --build .  && \
    cmake --build . --target install
RUN curl -Lsk https://dl.bintray.com/boostorg/release/1.71.0/source/boost_1_71_0.tar.bz2 | tar xj -C /opt
RUN cp /tmp/llvm-project-llvmorg-8.0.1/libcxxabi/include/* /opt/libcxx8_msan/include/
EOF
fi

if groups | grep docker; then
    DOCKERCMD=docker
else
    DOCKERCMD="sudo docker"
fi

DOCKERIMAGE=$($DOCKERCMD build environment_msan | tee /dev/tty | tail -n 1 | cut -d' ' -f3)
MOUNT=$(realpath mount)
if [ ! -d $MOUNT ]; then
    mkdir $MOUNT
fi
if [ ! -e $MOUNT/run ]; then
    cat <<'EOF' >$MOUNT/run.sh
#!/bin/bash -xe
export SYMENGINE_COMMIT=$1
if [ ! -d symengine-${SYMENGINE_COMMIT} ]; then
    wget --no-check-certificate https://github.com/symengine/symengine/archive/${SYMENGINE_COMMIT}.tar.gz
    tar xzf ${SYMENGINE_COMMIT}.tar.gz
fi
mkdir -p symengine-build
cd symengine-build
export CXXFLAGS="$CXXFLAGS -fsanitize=memory -stdlib=libc++ -I/opt/libcxx8_msan/include -I/opt/libcxx8_msan/include/c++/v1 -fno-omit-frame-pointer -fno-optimize-sibling-calls -O1 -glldb -DHAVE_GCC_ABI_DEMANGLE=no"
export LDFLAGS="-fsanitize=memory $LDFLAGS -Wl,-rpath,/opt/libcxx8_msan/lib -L/opt/libcxx8_msan/lib -lc++abi"
CXX=clang++-8 CC=clang-8 cmake \
    -DCMAKE_BUILD_TYPE:STRING="Debug" \
    -DWITH_BFD:BOOL=OFF \
    -DWITH_LLVM:BOOL=OFF \
    -DINTEGER_CLASS="boostmp" \
    -DWITH_SYMENGINE_RCP:BOOL=ON \
    -DBUILD_TESTS:BOOL=ON \
    -DBUILD_BENCHMARKS:BOOL=ON \
    -DBOOST_ROOT=/opt/boost_1_71_0 \
    ../symengine-${SYMENGINE_COMMIT}

make -j $(nproc)
ctest --output-on-failure
export MSAN_OPTIONS=symbolize=1,external_symbolizer_path=/usr/lib/llvm-8/bin/llvm-symbolizer
./symengine/utilities/matchpycpp/tests/bipartite/test_bipartite
./symengine/utilities/matchpycpp/tests/hopcroft_karp/test_hopcroft_karp
EOF
    chmod +x $MOUNT/run.sh
fi

HOST_USER=${SUDO_USER:-${LOGNAME}}
$DOCKERCMD run --rm --name "symengine_compile" \
       -e HOST_WHOAMI=${HOST_USER} -e HOST_UID=$(id -u ${HOST_USER}) -e HOST_GID=$(id -g ${HOST_USER}) \
       -v $MOUNT:/mount -w /mount --cap-add SYS_PTRACE \
       -it $DOCKERIMAGE /usr/bin/env bash -x -c \
       "groupadd -f --gid \$HOST_GID \$HOST_WHOAMI; \
useradd --uid \$HOST_UID --gid \$HOST_GID --home /mount \$HOST_WHOAMI; \
sudo --preserve-env --login -u \$HOST_WHOAMI ./run.sh $1"
```

</details>

**Comment by bjodah** [comment]
I updated the script to drop into gdb:
```diff
--- a/docker_symengine_msan.sh
+++ b/docker_symengine_msan.sh
@@ -35,6 +35,7 @@ RUN curl -Lsk https://github.com/llvm/llvm-project/archive/llvmorg-${LLVM_ORG_VE
     cmake --build . --target install
 RUN curl -Lsk https://dl.bintray.com/boostorg/release/1.71.0/source/boost_1_71_0.tar.bz2 | tar xj -C /opt
 RUN cp /tmp/llvm-project-llvmorg-8.0.1/libcxxabi/include/* /opt/libcxx8_msan/include/
+RUN apt-get update && apt-get --quiet --assume-yes --no-install-recommends install gdb
 EOF
 fi
 
@@ -51,7 +52,7 @@ if [ ! -d $MOUNT ]; then
 fi
 if [ ! -e $MOUNT/run ]; then
     cat <<'EOF' >$MOUNT/run.sh
-#!/bin/bash -xe
+#!/bin/bash -x
 export SYMENGINE_COMMIT=$1
 if [ ! -d symengine-${SYMENGINE_COMMIT} ]; then
     wget --no-check-certificate https://github.com/symengine/symengine/archive/${SYMENGINE_COMMIT}.tar.gz
@@ -75,8 +76,8 @@ CXX=clang++-8 CC=clang-8 cmake \
 make -j $(nproc)
 ctest --output-on-failure
 export MSAN_OPTIONS=symbolize=1,external_symbolizer_path=/usr/lib/llvm-8/bin/llvm-symbolizer
-./symengine/utilities/matchpycpp/tests/bipartite/test_bipartite
-./symengine/utilities/matchpycpp/tests/hopcroft_karp/test_hopcroft_karp
+gdb -ex "catch catch" -ex run -args ./symengine/utilities/matchpycpp/tests/bipartite/test_bipartite -x 1 -b
+gdb -ex "catch catch" -ex run -args ./symengine/utilities/matchpycpp/tests/hopcroft_karp/test_hopcroft_karp -x 1 -b
 EOF
     chmod +x $MOUNT/run.sh
 fi
@@ -88,4 +89,4 @@ $DOCKERCMD run --rm --name "symengine_compile" \
        -it $DOCKERIMAGE /usr/bin/env bash -x -c \
        "groupadd -f --gid \$HOST_GID \$HOST_WHOAMI; \
 useradd --uid \$HOST_UID --gid \$HOST_GID --home /mount \$HOST_WHOAMI; \
-sudo --preserve-env --login -u \$HOST_WHOAMI ./run.sh $1"
\ No newline at end of file
+sudo --preserve-env --login -u \$HOST_WHOAMI ./run.sh ${1:-d0b1932b4ffaf4b364d4b436a796b456917e079b}"
```

Still I find it hard to interpret the backtrace. I'm wondering if msan output is getting redirected when run under catch/ctest.

**Comment by bjodah** [comment]
I updated the stand-alone script to reproduce the memory sanitizer errors in `test_bipartite` and `test_hopcroft_karp`:
https://gist.github.com/bjodah/82e1013e2025003e35f365b89921ddc2

Looks like those two tests fail consistently.

So I think this PR works as is. Would be nice to hunt down the bugs (or confirm that we're looking at false positives) in matchpycpp at some point, but I won't have the time to do so any time soon. In the mean time, I think adding this to the test suite could be good to protect against introducing any new errors of this class to the code base?

**Comment by certik** [comment]
I agree, this looks great.

One issue that I can see is that the `fold_start` effectively drops the lines in between, there does not seem to be a way to actually see the lines. So I would remove it and always show the full output.

**Comment by bjodah** [comment]
Sure (I could have sworn it was working at an earlier point, doesn't matter, I'll just drop those lines).

**Comment by bjodah** [comment]
Turns out that the use of catch's SECTION has issues when using libc++:
https://github.com/catchorg/Catch2/issues/1214
https://github.com/catchorg/Catch2/issues/1218

Hence, the failures have nothing to do with memory sanitizer, but with the fact we're using libc++, so I think it's OK to skip those tests for now (would be nice to be able to run all tests when using libc++, but that's a separate issue).

So from my point of view this is good to go.

**Pull request merged**